### ISSUE TITLE
How to describe a column as a `non-empty-string` ?

### ISSUE THREAD

**Issue by Crovitche-1623** [phpstan/phpstan-doctrine#635]
Since Symfony 7.2, the method `getUserIdentifier()` from `UserInterface.php` interface requires to return a `non-empty-string`. However, why I try to describe my property as a non-empty-string, PHPStan report this:
```
Property App\Entity\Account::$username type mapping mismatch: database can contain string but property expects non-empty-string|null.
```

for the following configuration:
```php
    /**
     * @var  non-empty-string|null  $username
     */
    private ?string $username = null;
```

I know entities need to be sometimes in an "invalid state" because we use the same PHP objects for Symfony Forms. However, when entities are persisted in the database, we know they are validated.

@ondrejmirtes What is the recommended way to treat this warning. Create a custom doctrine type for non-empty-string ?

**Comment by ondrejmirtes** [comment]
ORM column string mapping type can't have minLength, right? Custom type it is then.

**Comment by Crovitche-1623** [comment]
Thank you for the quick answer ! Indeed, the ORM string mapping type does not have a built-in `minLength` parameter. I'll create a custom type then.

**Comment by Crovitche-1623** [comment]
Here is my working solution if anybody is interested:

The custom Doctrine data type: 
```php
<?php

declare(strict_types=1);

namespace App\Doctrine\Type;

use Doctrine\DBAL\Platforms\AbstractPlatform;
use Doctrine\DBAL\Types\Exception\InvalidType;
use Doctrine\DBAL\Types\StringType;

/**
 * A workaround to type-hint a column with a `non-empty-string` datatype.
 *
 * @author  Thibault Gattolliat  <contact@thibaultg.info>
 */
class NonEmptyStringType extends StringType
{
    /**
     * @var  non-empty-string
     */
    public const NAME = 'non-empty-string';

    public function getName(): string
    {
        return self::NAME;
    }

    /**
     * {@inheritDoc}
     *
     * @return  non-empty-string|null
     */
    #[\Override]
    public function convertToDatabaseValue(
        mixed $value,
        AbstractPlatform $platform,
    ): ?string {
        if (null === $value) {
            return null;
        }

        if (!\is_string($value) || '' === $value)) {
            throw InvalidType::new(
                $value,
                static::class,
                ['null', self::NAME],
            );
        }

        /**
         * @var  non-empty-string  $return
         */
        $return = parent::convertToDatabaseValue($value, $platform);

        return $return;
    }

    /**
     * {@inheritDoc}
     *
     * @return  non-empty-string|null
     */
    #[\Override]
    public function convertToPHPValue(
        mixed $value,
        AbstractPlatform $platform,
    ): ?string {
        if (null === $value) {
            return null;
        }

        if (!\is_string($value) || '' === $value)) {
            throw InvalidType::new(
                $value,
                static::class,
                ['null', self::NAME],
            );
        }

        return $value;
    }
}
```

The PHPStan descriptor:
```php
<?php

declare(strict_types=1);

namespace App\PHPStan\Type\Doctrine\Descriptors;

use App\Doctrine\Type\NonEmptyStringType;
use PHPStan\Type\Accessory\AccessoryNonEmptyStringType;
use PHPStan\Type\Doctrine\Descriptors\DoctrineTypeDescriptor;
use PHPStan\Type\Type;

/**
 * This class allows PHPStan to describe a `non-empty-string` datatype.
 *
 * @author  Thibault Gattolliat  <contact@thibaultg.info>
 */
final class NonEmptyStringDescriptor implements DoctrineTypeDescriptor
{
    public function getType(): string
    {
        return NonEmptyStringType::class;
    }

    public function getWritableToPropertyType(): Type
    {
        return new AccessoryNonEmptyStringType();
    }

    public function getWritableToDatabaseType(): Type
    {
        return new AccessoryNonEmptyStringType();
    }

    public function getDatabaseInternalType(): Type
    {
        return new AccessoryNonEmptyStringType();
    }
}
```

Configuration for Doctrine ORM:
```yaml
doctrine:
    dbal:
        types:
            !php/const App\Doctrine\Type\NonEmptyStringType::NAME: App\Doctrine\Type\NonEmptyStringType
```

Usage:
```php
<?php

declare(strict_types=1);

namespace App\Entity;

use App\Doctrine\Type\NonEmptyStringType;
use Doctrine\ORM\Mapping as ORM;

// ...
class FooBar {

    /**
     * @var  non-empty-string|null  $username
     */
    #[ORM\Column(
        type: NonEmptyStringType::NAME,
    )]
    private ?string $username = null;
}
```

This solution could be improved using a custom `minLength` parameter.

**Comment by ondrejmirtes** [comment]
Just be careful, empty() returns true for `'0'`. So in fact your type produces non-falsy-string, not just non-empty-string.

**Comment by Crovitche-1623** [comment]
@ondrejmirtes Indeed, I confused the two types. I corrected my example to match the `non-empty-string` instead.
